Core/src/Material/IntersectionMaterialAssigner.cpp

```diff
@@ -0,0 +1,116 @@
+    const Acts::GeometryContext& gctx, const Acts::Vector3& position,
+    const Acts::Vector3& direction,
+    const std::vector<const Acts::Surface*>& surfaces) {
+  // First deal with the surface intersections
+  std::vector<Acts::SurfaceIntersection> sIntersections;
+  // Intersect the surfaces
+  for (auto& surface : surfaces) {
+    // Get the intersection
+    auto sMultiIntersection = surface->intersect(gctx, position, direction,
+                                                 Acts::BoundaryCheck(true));
+    // One solution, take it
+    if (sMultiIntersection.size() == 1u &&
+        sMultiIntersection[0u].status() >=
+            Acts::IntersectionStatus::reachable &&
+        sMultiIntersection[0u].pathLength() >= 0.0) {
+      sIntersections.push_back(sMultiIntersection[0u]);
+      continue;
+    }
+    if (sMultiIntersection.size() > 1u) {
+      // Multiple intersections, take the closest
+      auto closestForward = sMultiIntersection.closestForward();
+      if (closestForward.status() >= Acts::IntersectionStatus::reachable &&
+          closestForward.pathLength() > 0.0) {
+        sIntersections.push_back(closestForward);
+        continue;
+      }
+    }
+  }
+  // Sort the intersection along the pathlength
+  std::sort(sIntersections.begin(), sIntersections.end(),
+            &Acts::SurfaceIntersection::pathLengthOrder);
+  return sIntersections;
+}
+
+}  // namespace
+
+std::pair<std::vector<Acts::IAssignmentFinder::SurfaceAssignment>,
+          std::vector<Acts::IAssignmentFinder::VolumeAssignment>>
+Acts::IntersectionMaterialAssigner::assignmentCandidates(
+    const GeometryContext& gctx, const MagneticFieldContext& /*unused*/,
+    const Vector3& position, const Vector3& direction) const {
+  // The resulting candidates
+  std::pair<std::vector<Acts::IAssignmentFinder::SurfaceAssignment>,
+            std::vector<Acts::IAssignmentFinder::VolumeAssignment>>
+      candidates;
+
+  // Try the surfaces first
+  auto sIntersections =
+      forwardOrderedIntersections(gctx, position, direction, m_cfg.surfaces);
+  candidates.first.reserve(sIntersections.size());
+  for (auto& sIntersection : sIntersections) {
+    candidates.first.push_back(std::make_tuple(
+        sIntersection.object(), sIntersection.position(), direction));
+  }
+
+  // Now deal with the volume intersections : tracking volume first
+  if (!m_cfg.trackingVolumes.empty()) {
+    for (auto& trackingVolume : m_cfg.trackingVolumes) {
+      // Collect the boundary surfaces
+      auto boundarySurfaces = trackingVolume->boundarySurfaces();
+      std::vector<const Surface*> tSurfaces;
+      for (auto& boundarySurface : boundarySurfaces) {
+        tSurfaces.push_back(&(boundarySurface->surfaceRepresentation()));
+      }
+      // Get the intersections
+      auto tIntersections =
+          forwardOrderedIntersections(gctx, position, direction, tSurfaces);
+      // Entry/exit exists in forward direction
+      if (tIntersections.size() == 2u) {
```

Comment:
`!= 2` can be ignored? or should it be an error case?

Reply:
This code is not yet tested, I will add tests and then refine the logic here.